provider/packet/provider.go

```diff
@@ -0,0 +1,137 @@
+)
+
+type environProvider struct {
+	environProviderCredentials
+}
+
+func (p environProvider) CloudSchema() *jsonschema.Schema {
+	trace()
+	return nil
+}
+
+// Ping tests the connection to the cloud, to verify the endpoint is valid.
+func (p environProvider) Ping(ctx context.ProviderCallContext, endpoint string) error {
+	trace()
+	return errors.NotImplementedf("Ping")
+}
+
+// PrepareConfig is part of the EnvironProvider interface.
+func (p environProvider) PrepareConfig(args environs.PrepareConfigParams) (*config.Config, error) {
+	trace()
+	if err := validateCloudSpec(args.Cloud); err != nil {
+		return nil, errors.Annotate(err, "validating cloud spec")
+	}
+	// Set the default block-storage source.
+	attrs := make(map[string]interface{})
+	// if _, ok := args.Config.StorageDefaultBlockSource(); !ok {
+	// 	attrs[config.StorageDefaultBlockSourceKey] = azureStorageProviderType
+	// }
+	// fmt.Println("config attrs", args.Config)
+
+	if len(attrs) == 0 {
+		return args.Config, nil
+	}
+	return args.Config, nil
+}
+
+func validateCloudSpec(spec environscloudspec.CloudSpec) error {
+	trace()
+	// if err := spec.Validate(); err != nil {
+	// 	return errors.Trace(err)
+	// }
+	// if spec.Credential == nil {
+	// 	return errors.NotValidf("missing credential")
+	// }
+	// if authType := spec.Credential.AuthType(); authType != clientCredentialsAuthType {
+	// 	return errors.NotSupportedf("%q auth-type", authType)
+	// }
+	return nil
+}
+
+// Open is specified in the EnvironProvider interface.
+func (p environProvider) Open(args environs.OpenParams) (environs.Environ, error) {
+	trace()
+	logger.Debugf("opening model %q", args.Config.Name())
+
+	e := new(environ)
+	e.name = args.Config.Name()
+
+	if err := e.SetCloudSpec(args.Cloud); err != nil {
+		return nil, err
+	}
+
+	if err := e.SetConfig(args.Config); err != nil {
+		return nil, errors.Trace(err)
+	}
+	return e, nil
+}
+
+func (environProvider) Validate(cfg, old *config.Config) (valid *config.Config, err error) {
+	trace()
+	newEcfg, err := validateConfig(cfg, old)
+	if err != nil {
+		return nil, fmt.Errorf("invalid Packet provider config: %v", err)
+	}
+	return newEcfg.config.Apply(newEcfg.attrs)
+}
+
+func (p environProvider) newConfig(cfg *config.Config) (*environConfig, error) {
+	trace()
+	valid, err := p.Validate(cfg, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &environConfig{valid, valid.UnknownAttrs()}, nil
+}
+
+func (e *environ) SetCloudSpec(spec environscloudspec.CloudSpec) error {
+	trace()
+	e.ecfgMutex.Lock()
+	defer e.ecfgMutex.Unlock()
+
+	e.cloud = spec
+
+	// var err error
+	e.packetClient = packetClient(e.cloud)
+
+	return nil
+}
+
+func packetClient(spec environscloudspec.CloudSpec) *packngo.Client {
+	trace()
+	credentialAttrs := spec.Credential.Attributes()
+
+	apiToken := credentialAttrs["api-token"]
+	httpClient := retryablehttp.NewClient()
+	httpClient.RetryWaitMin = time.Second
+	httpClient.RetryWaitMax = 30 * time.Second
+	httpClient.RetryMax = 10
```

Comment:
This seems a bit liberal, essentially we're waiting for 5m10s for every request if it fails to connect to the server? I'm a fan of retries, but we should be specific about what retries we want to do, rather than just retry everything...

Reply:
The problem is with packngo
```
func NewClientWithAuth(consumerToken string, apiKey string, httpClient *retryablehttp.Client) *Client {
```
which is expecting to receive this type of httpClient.
Probably it would be a good idea to refactor this in the Packet SDK.

Reply:
Is there any particular reason for requiring a concrete `retryablehttp.Client` instance there rather an interface for performing http requests?

Reply:
There were cases where the API was returning empty responses and was recovering after that. We would need to confirm that this isn't the case anymore. Can we confirm that @displague?

Reply:
What I mean is that if the retry logic is embedded in the client and the packet client simply calls `Do` or `Get` on it, you could change the signature to accept an interface with the methods you need and then consumers can pass either the retryablehttp client or any generic client (like the one that juju uses).